README.md
# Sarge AI System
This is an AI system that is intended to simulate real players roaming the map, scavenging, looting and engaging in combat. Unlike other AI systems Sarge AI designates two factions as friendly and only one faction as hostile. This approach provides a simulated pvp/pve environment that changes depending on how players react to the AI.

## Behaviour
AI hostility is based on player rating and AI will change their behaviour depending on a players rating level. Above the threshold the friendly AI will not attack players. Below the threshold friendly AI will consider the player hostile and fire on sight. Bandit AI will always be hostile to players regardless of player rating.

## Customizing
Almost every setting you can think of is available to be customized to each server owners liking. This includes weapons, uniforms, vests, headgear, cars, helicopters and much more. Server owners even have the option of creating their own map markers for static AI to patrol, fortify, ambush and other tactical approaches.

## Units
* Survivors
  * Fighting for survival as most of us are; These AI do not wish to start unneccessary confilict as they intend on living as much as any player. These AI will assume every player is friendly unless they are given reason to think otherwise. They often wear toned down military apparel and civilian clothing and should be easy to spot as to not frighten unaware players.

* Military
  * As the dominant force in a wasteland of fugitives and other life threatening obstacles the military takes no chance when it comes to combat. Dressed in high level miliatry apparel and the weapons to back it up these NPCs will roam about the area they spawn in and nuetralize any threat they encounter. Dont' be a threat ;)

* Bandits
  * With no apparent loyalty to anyone but the small group of others they travel with these NPCs will roam the map surviving any way they are able to. These AI will steal any loot they encounter and attempt to kill any player they encounter. Outfitted with respectable weapons and headwraps to protect their identities from the authorities these AI make for a formidable enemy. 

This description is brief but to the point. Although it may not be listed you will find these NPCs are capable of suprising things.


changeLog.md
#### 2.4.5
- [Fix] Changed bandit unit model to Opfor unit to fix issue causing bandits not to be hostile.

#### 2.4.4
- [New] Added support for the BreakingPoint mod.

#### 2.4.3
- [Fix] Fixed a bug that prevented Epoch players from entering vehicles.

#### 2.4.2
- [Fix] Fixed undefined variables for AI vehicles.

#### 2.4.0
- [New] Vests have been added to the customization options.
- [New] Backpacks have been added to the customization options.
- [New] Sarge AI now supports Desolation, Epoch and Exile mods and is determined by the class names of CfgMods.
- [Change] All configuration files are now in the "config" folder.
- [Change] Error messages have been redone to be more specific about what went wrong when spawning AI and generating gear.
- [Change] All chernarus maps will be processed separately for now until I can work out a more reliable way of using one map file for all variants.
- [Change] All map files have been formatted with the proper logic to properly handle dynamic and static calls.
- [Change] AI now have unique voices per type and should be able to speak to other AI group members allowing players to hear the AI.
- [Fix] Optional UPSMON actions will now be forced to uppercase instead of lower case to be more aligned with UPSMON.
- [Fix] An issue causing the uniform argument to be passed as an array on helicopter patrol units has been fixed.
- [Fix] A count check has been added to all loadout arrays before processing them to avoid errors when processing empty arrays.
- [Fix] AI not being given a speaker has been resolved and should no longer spam rpt logs when spawning new AI.
- [Bug] Static spawns are not working as expected and have been disabled for now.
- [Bug] Headless client has been disabled to begin work on a more appropriate method.

#### 2.3.1
- [New] Vests have been added to the customization options.
- [New] Backpacks have been added to the customization options.
- [Change] Error messages have been redone to be more specific about what went wrong when spawning AI and generating gear.
- [Change] Headless client will be detected automatically and used at random if one or more exists.
- [Change] All chernarus maps will be processed separately for now until I can work out a more reliable way of using one map file for all variants.
- [Change] All map files have been formatted with the proper logic to properly handle dynamic and static calls.
- [Change] AI now have unique voices per type and should be able to speak to other AI group members allowing players to hear the AI.
- [Fix] Optional UPSMON actions will now be forced to uppercase instead of lower case to be more aligned with UPSMON.
- [Fix] An issue causing the uniform argument to be passed as an array on helicopter patrol units has been fixed.
- [Fix] A count check has been added to all loadout arrays before processing them to avoid errors when processing empty arrays.
- [Fix] AI not being given a speaker has been resolved and should no longer spam rpt logs when spawning new AI.

#### 2.2.9
- [Fix] Fixed a bug causing friendly AI to have head wraps.
- [New] AI women have been added to the spawn function.
- [Change] AI Skills decreased to 10% multiplied by a maximum of 2 bringing the highest possible skill level to 20%.
- [Change] AI loadout variable names were changed to a more uniform template to simplify the customization process.
- [Change] Major changes to code blocks and variables were done to make customizing more intuitive for server owners.
- [Change] AI are tied to nuisance, rewarding players an +10 points for bandit kills and -20 points for friendly kills.

Known Issues
	- Error Undefined variable in expression: upsmon_guer_total

#### 2.2.8
- [Change] UPSMON scripts are now located in the mission PBO as was intended by the creator to remove instability issues with AI logic.

#### 2.2.3
- [Fix] Static infantry now spawn as intended.
- [Change] UPSMON scripts have been merged with the Sarge PBO.

#### 2.2.2
- [Fix] Hotfix for AI stealing vehicles.

#### 2.2.1
- [Change] Sarge AI and UPSMON have been converted to an addon format.

#### 2.1.7
- Sarge AI is now in transition to becoming an official addon. The Sarge AI code has been converted and the UPSMON may be converted slowly or replaced completely.
- The installation instructions have changed drastically compared to the mission PBO version.

#### 2.1.6
- If you are using a headless client and you install the development version you MUST REMOVE ELEC_DETEC VERSION COMPLETELY.
- This method will automatically detect where the headless client is and use it without and configurations needed.

Headless Client
- HC logic has been reverted back to a more Arma friendly method despite a bug with bandit AI groups.
- Bandits within groups will sometimes spawn as a different side than intended and will be killed by non bugged badits.
- Bandit max group count has been doubled to offset the effect of this HC bandit bug.

Dynamic Heli Spawns
- Huge failure! LOL, put on hold for better development.

Experience System
- AI will now start at level 1 and level up depending on individual experience ratings.
- Working on restoring damage reduction/additions based on AI level.

#### 2.1.5 Changes
- [Change] Improved HC spawning logic to provide a better HC experience.
- [New] AI backpacks have been added and should no longer spawn with items.
- [New] AI will now have a sidearm in addition to their main weapon.
- [New] Added more clothing options to all AI to improve aesthetics.
- [New] Added a variable to control the Sarge AI system chat messages.
- [New] Added a variable to control the AI's ability to speak and use side chat.
- [New] Added a persistent variable to track friendly and hostile kills.
- [Fix] Fixed a bug that caused some friendly AI to appear as a bandit.

#### 2.1.0 Changes
- [Removed] The experience system for the AI has been removed.

sarge/SAR_define.hpp
/*
	# Original #
	Sarge AI System 1.5
	Created for Arma 2: DayZ Mod
	Author: Sarge
	https://github.com/Swiss-Sarge

	# Fork #
	Sarge AI System 2.0+
	Modded for Arma 3: Exile Mod
	Changes: Dango
	https://www.hod-servers.com

*/
class CfgIdentities {
	class id_SAR {
		name = "id_SAR";
        face = "WhiteHead_06";
		glasses = "None";
		speaker = "NoVoice";
		pitch = 1.00;
	};
	class id_SAR_band : id_SAR {
		name = "id_SAR_band";
		face = "PersianHead_A3_02";
	};
	class id_SAR_sold_lead : id_SAR {
		name = "id_SAR_sold_lead";
		face = "WhiteHead_02";
	};
	class id_SAR_sold_man : id_SAR {
		name = "id_SAR_sold_man";
		face = "WhiteHead_02";
	};
	class id_SAR_surv_lead : id_SAR {
		name = "id_SAR_surv_lead";
		face = "WhiteHead_02";
	};
	class id_SAR_surv_man : id_SAR {
		name = "id_SAR_surv_man";
		face = "WhiteHead_04";
	};
};

sarge/config.cpp
/*
	# Original #
	Sarge AI System 1.5
	Created for Arma 2: DayZ Mod
	Author: Sarge
	https://github.com/Swiss-Sarge

	# Fork #
	Sarge AI System 2.0+
	Modded for Arma 3: Exile Mod
	Changes: Dango
	http://www.hod-servers.com
	https://github.com/Teh-Dango
	
*/
class CfgPatches {
	class Sarge_AI {
		requiredVersion = 1.36;
		version = 2.4;
		requiredAddons[] = {};
		units[] = {};
		weapons[] = {};
		magazines[] = {};
		ammo[] = {};
	};
};

class CfgFunctions {
	class SAR {
		class main {
			file = "addons\sarge\init";
			class preInit {
				preInit = 1;
			};
			class postInit {
				postInit = 1;
			};
		};
		class compiles {
			file = "addons\sarge\code\functions";
			class AI_anim_heli {};
			class AI_despawn {};
			class AI_guards {};
			class AI_heli {};
			class AI_hit {};
			class AI_hit_vehicle {};
			class AI_infantry {};
			class AI_interact {};
			class AI_killed {};
			class AI_refresh {};
			class AI_spawn {};
			class AI_trace {};
			class AI_trace_base {};
			class AI_trace_vehicle {};
			class AI_traders {};
			class AI_vehicle {};
		};
	};
};

class CfgIdentities {
	class id_SAR {
		name = "id_SAR";
        face = "WhiteHead_05";
		glasses = "None";
		speaker = "Male05ENG";
		pitch = 1.00;
	};
	class id_SAR_band : id_SAR {
		name = "id_SAR_band";
		face = "PersianHead_A3_01";
		speaker = "Male01PER";
	};
	class id_SAR_sold_lead : id_SAR {
		name = "id_SAR_sold_lead";
		face = "WhiteHead_02";
		speaker = "Male05ENG";
	};
	class id_SAR_sold_man : id_SAR {
		name = "id_SAR_sold_man";
		face = "WhiteHead_03";
		speaker = "Male06ENG";
	};
	class id_SAR_surv_lead : id_SAR {
		name = "id_SAR_surv_lead";
		face = "WhiteHead_04";
		speaker = "Male07ENG";
	};
	class id_SAR_surv_man : id_SAR {
		name = "id_SAR_surv_man";
		face = "WhiteHead_05";
		speaker = "Male08ENG";
	};
};